backupccl: pause scheduled backup if resumed on different cluster

msbutler opened the pull request:
Previously, after cluster restore or c2c cutover, a backed up/ replicated
backup schedule would begin executing on the new cluster. If the backup /
source cluster was still available and executing the schedule, then the two
schedules would compete to run, as outlined in https://github.com/cockroachdb/cockroach/issues/108028.

This patch prevents this poor UX by pausing the backup schedule if the schedule
realizes it is running on a new cluster. It is then up to the user to resume
backups on the new cluster and prevent the backup collision problem.

Fixes https://github.com/cockroachdb/cockroach/issues/108028

Release note (sql change): if a scheduled backup resumes on a new cluster (e.g.
after C2C cutover or a tenant restore), the backup schedule will pause. The
user may resume the schedule without changing it, but should take special care
to ensure no other schedule is backing up to the same collection. The user may
also want to cancel the paused schedule and start a new one.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111578)
<!-- Reviewable:end -->

msbutler commented:
@dt @adityamaru This is ready for a look. The first two commits are getting reviewed in other prs: Test server change is [here](https://github.com/cockroachdb/cockroach/pull/111603), and the scheduled backup change[ here](https://github.com/cockroachdb/cockroach/pull/111374). 

I have one high level design q: should I check for the clusterID in the `scheduledBackupExecutor` implementation of `ExecuteJob` (current approach), or should I add a new public method to the `ScheduledJobExecutor` interface along the lines of `CheckForNewCluster()`?

dt commented on pkg/ccl/backupccl/schedule_exec.go:
```diff
@@ -99,19 +99,33 @@ func (e *scheduledBackupExecutor) executeBackup(
+	currentClusterID := planner.ExtendedEvalContext().ClusterID
+	currentDetails := sj.ScheduleDetails()
+	if currentClusterID != currentDetails.ClusterID {
+		log.Infof(ctx, "pausing backups Schedule %d and updating Cluster ID in schedule details. persisted cluster ID %d is different than current cluster ID %d",
```
nit: I'd flip and shorten this; "schedule last run by cluster %s, pausing until manually resumed"

dt commented on pkg/ccl/backupccl/schedule_exec.go:
```diff
@@ -99,19 +99,33 @@ func (e *scheduledBackupExecutor) executeBackup(
+	planner := hook.(sql.PlanHookState)
+	currentClusterID := planner.ExtendedEvalContext().ClusterID
+	currentDetails := sj.ScheduleDetails()
+	if currentClusterID != currentDetails.ClusterID {
```
and currentDetails.ClusterID non-empty right? for backwards compatibility?

msbutler replied:
oooof. nice freaking catch. fixed.

dt reviewed (commented):
2¢: firing up a whole c2c end to end test with two tenants and cutover and schedules all involved feels like a kinda expensive way to cover this one behavior of schedules to me, and far away from the code in question. I might vote to just test this by manually creating a schedule with a different (and empty!) creator in create_scheduled_backup_test.go, removing all streamingccl/ diffs?

msbutler committed:
backupccl: pause scheduled backup if resumed on different cluster

Previously, after cluster restore or c2c cutover, a backed up/ replicated
backup schedule would begin executing on the new cluster. If the backup /
source cluster was still available and executing the schedule, then the two
schedules would compete to run, as outlined in #108028.

This patch prevents this poor UX by pausing the backup schedule if the schedule
realizes it is running on a new cluster. It is then up to the user to resume
backups on the new cluster and prevent the backup collision problem.

Fixes #108028

Release note (sql change): if a scheduled backup resumes on a new cluster (e.g.
after C2C cutover or a tenant restore), the backup schedule will pause. The
user may resume the schedule without changing it, but should take special care
to ensure no other schedule is backing up to the same collection. The user may
also want to cancel the paused schedule and start a new one.

msbutler commented:
@dt ready for another look! Added a better test in create_scheduled_backup_test.go

msbutler commented:
bors r=dt
